### ISSUE TITLE
Using your code, questions 

### ISSUE THREAD

**Issue by dyhan316** [gjoelbye/SPEED#2]
Hello! I am interested in using your work to preprocess TUEG and our custom ECoG data, and was wondering if you could shed light on some questions : 

1. What is the proportion of data that is rejected when you apply the QA? Is it significant? 
1. **montages** : it seems that your pipeline requires a montage to be set before running. Is there a way to (a) : set a custom montage (not just choose from 'tuh' or 'standard_1020' (b) : Can I make it so that all channels are preserved (and not dropped for being not in the montage)
2. Is there a way to skip channel interpolation? (And just make the missing channel data be NaN or zero?)
3. If I don't want the final data to be windowed, is it OK for me to just concat all the windows? (i.e. when running TUEG the resulting h5 data is of shape `data.shape = (21, 19, 15360)`. Can I just do `data.resahpe(19,-1)`? 
7. I have a question about the average referencing you use. If I understand correctly, TUEG has two reference schemes : AR (average referencing) and LE (linked ear). Since you rereferecne AFTER you take out channels not in the montage, the channels you use for rereferecing and the channels that were used for initial refereeing (when TUEG gave it to us) will be different, and hence proper average referencing might not be done?
8. Is it possible to rereference to LE scheme? Is this worth it you think?
9. When applying line noise, should I do [60, 120, 180, ..] to remove harmonics?
9. Do you have thoughts on expanding to SEED and other datasets?

Thank you in advance for your answer! Sorry for asking too much

**Comment by gjoelbye** [comment]
Hi Dyhan! Thank you for your interest and thoughtful questions! I'm continuously working to enhance the pipeline, and I'll address your points below:

1. Proportion of Data Rejected: This varies by dataset. For TUH, around 5–10% of data is typically dropped due to quality issues. For smaller, curated datasets, rejection rates are much lower. You can adjust thresholds to reduce rejections further.

2. Montages: The pipeline supports all MNE montages. To add custom montages, you can follow the approach used for the TUH montage—create a function in src/utils.py that returns your montage. You can adapt existing MNE montages, combine them, or load new montage files (text files with channel names and locations). Currently, preserving all channels isn’t directly supported, as methods like ICA and interpolation depend on known channel locations.

3. Skipping Channel Interpolation: While direct skipping isn't supported, you can mix elements from the SPEED pipeline (which interpolates) with the DynamicPipelineBENDR pipeline (which zeroes out missing channels) to achieve a similar result.

4. Concatenating Non-Windowed Data: This approach could work but be cautious not to concatenate non-neighboring windows if some were dropped. Alternatively, consider the PyPrep pipeline, which might align with your needs.
 
5. Average Referencing Concerns: You're correct that rereferencing after dropping bad channels could differ from the original TUEG references. We chose this method to avoid corrupting the average reference with bad channels. Using all channels, even those outside the montage, for average referencing could work but requires a different approach that doesn't rely on channel location knowledge.

6. Linked Ear Referencing: I do not know if it is possible but it sounds technically challenging for most EEG datasets. For machine learning applications, using linked ear referencing may not significantly impact performance.

7. Line Noise Removal: Yes, harmonics (e.g., 60 Hz, 120 Hz, 180 Hz) are removed. The SPEED pipeline already handles this using the Zapline method, which handles harmonics effectively.

8. Expanding to SEED and Other Datasets: The pipeline is designed to be versatile, but EEG data comes in many formats. The next step is likely to expand support and improving downstream compatibility.

I hope this helps! Best of luck with your project!

**Comment by dyhan316** [comment]
Thank you so much @gjoelbye  for your detailed response!

Actually I addressed most of the questions myself by customizing your code a lot, but still thank you for the answer. Also thank you for the code, it's very useful!

I have some follow up questions if you don't mind : 
1. -> Is there a specific threshold value (oha, thv, chv values) that you can recommend for looser rejection criteria? (Or have you already tried some values?) 
2. -> Thank you so much for the info! I'll look into it :)  
3. -> thank you 
4. -> thank you! we decided not to do it as 60s seems more than long enough for pretraining
5. -> thank you
6. -> thank you 
7. -> Thank you! so to clarify, I should do [60,120, ...] in the yaml file right?
8. -> thank you 

========
I also have a few more new questions if you don't mind : 

1. We are thinking preprocessing private ECoG dataset.  : **Does the QA pipeline (ex : bad channel rejection and such) also work for iEEG datasets too? Or is there another pipeline that should be used/implemented?**



Also, here are the added options that I have implemented. If you are interested, I could share the code! (but code is very dirty haha)(Other opinions are also welcome!) 
```python
#! ADDED PARAMETERS FOR US 
keep_remainder : bool = False, #if True, will keep the remainder of the data after splitting (last window will be longer due to the added thing)
run_ransac : bool = True, #if False, will not do ransac during QA
run_detrend : bool = True, #if False, will not detrend during preprocessing (meegkit detrending)
preproc_lp_freq : Union[float, str] = "same", #(during preproc) if None : no high pass, if 'same', will use the same lp_freq for QA and preprocessing.  
preproc_hp_freq : Union[float, str] = "same", #(during preproc) if None, no low pass, if 'same', will use the same hp_freq for QA and preprocessing
preproc_line_noise_option : Union[float, str] = "default" , #if default, will use the default line noise removal option, if mnenotch, use MNE's default, if False/None, will not remove line noise
preproc_line_freqs : Union[List[float], None, str] = "same", #(during preproc) if same, will use the same line_freqs for QA and preprocessing
rereference_option : Union[str, bool] = 'ar', #what rereference option to use for preprocessing. ar, le, or (None or False) (no rereferencing)
missing_chs_option : str = 'interpolate', #what to do with missing channels. 1. interpolate (default), 2. zero, 3. remove (remove window), or 4. ignore (keep on despite the missing channels)
drop_extra_chs : bool = True, #if False, will not drop extra channels
reorder_chs : bool = True, #if False, will not reorder channels
resample_method : Union[str, bool] = 'default', #options : 'default' (SPEED's interpolate_nearest), 'mne' (MNE's interpolate), 'decimate' (not implemented yet), False/None (no resampling)
```

**Comment by dyhan316** [comment]
I have an additional question if you don't mind 


in `methods.py`, when you do resampling (inteprolate_nearest)(code below)
```
    def interpolate_nearest(raw, sfreq=256.0):
        x = raw._data

        old_sfreq = raw.info['sfreq']
        
        resampled_data = interpolate(tensor(x).unsqueeze(0), 
                                     scale_factor=sfreq/old_sfreq, 
                                     mode="nearest").squeeze(0).numpy()
        
        lowpass = raw.info.get("lowpass")
        with raw.info._unlock():
            raw.info["sfreq"] = sfreq
            raw.info["lowpass"] = min(lowpass, sfreq / 2.0)
            
        raw._data = resampled_data
        raw._last_samps = np.array([resampled_data.shape[1] - 1])
```


you update the info of the raw with `lowpass`. I believe this is done to notify the fact that lowpass was applied. 

However, the `interpolate` function you use (`from torch.nn.functional import interpolate`) by default does not do anti-aliasing, and hence the lowpass should not be changed. (https://pytorch.org/docs/stable/generated/torch.nn.functional.interpolate.html) 

is my observation correct? 

Thank you again!
